### DISCUSSION TITLE
[PSD12] Exercise 3.26

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#185] [Esercizi (A-L)]
**Exercise 3.26**

You have been enlisted to design a soda machine dispenser for your department lounge. Sodas are partially subsidized by the student chapter of the IEEE, so they cost only 25 cents. The machine accepts nickels (5 cents), dimes (10 cents), and quarters (25 cents). When enough coins have been inserted, it dispenses the soda and returns any necessary change. Design an FSM controller for the soda machine. The FSM inputs are _Nickel_, _Dime_, and _Quarter_, indicating which coin was inserted. Assume that exactly one coin is inserted on each cycle. The outputs are _Dispense_, _ReturnNickel_, _ReturnDime_, and _ReturnTwoDimes_. When the FSM reaches 25 cents, it asserts _Dispense_ and the necessary _Return_ outputs required to deliver the appropriate change. Then it should be ready to start accepting coins for another soda.

**Comment by Elia-Belli** [comment]
![S1](https://user-images.githubusercontent.com/107944476/212889198-4a7866c3-d9c6-4ba6-82c6-053012e6bb5f.jpg)
![S2](https://user-images.githubusercontent.com/107944476/212889221-7125fec7-0965-4e02-824f-bb671f00edb8.jpg)

**Reply by ojaskudari** [reply]
```
module Soda(
    input clk,rst_n,
    input  Nickel, Dime,  Quarter,
    output reg Dispsense,
    output reg ReturnNickel, ReturnDime, ReturnTwoDimes
);

reg[3:0] PS,NS;
parameter [1:0] S0=2'd0,S1=2'd1,S2=2'd2;
reg [4:0]Cost,Change;

always@(posedge clk or negedge rst_n) begin
    if(!rst_n)
        PS <= S0;
    else
        PS <= NS;
end

always@(posedge clk)
begin
    case(PS)
    S0: 
    begin
        Dispsense <= 0;
        ReturnNickel <= 0;
        ReturnDime <= 0;
        ReturnTwoDimes <= 0;
        Cost<=0;
    end

    S1:
    begin
        if(Quarter)
        Cost<=Cost+5'd25;
        if(Dime)
        Cost<=Cost+5'd10;
        if(Nickel)
        Cost<=Cost+5'd5;
    end
    S2:
    begin
        Cost<=5'd0;
        Change<=Cost-5'd25;
    end
    endcase
end

always@(*)
begin
    NS=PS;
    case(PS)
    S0:NS=(Nickel|Dime|Quarter)?S1:S0;
    S1:NS=(Cost>=5'd25)?S2:S1;
    S2:NS=S0;
    endcase

end

always@(*)
begin
    if(PS==S2)
    begin
        Dispsense=1;
        ReturnNickel=(Change==5'd5)?1:0;
        ReturnDime=(Change==5'd10)?1:0;
        ReturnTwoDimes=(Change==5'd20)?1:0;
    end
    else
    begin
        Dispsense=0;
        ReturnDime=0;
        ReturnNickel=0;
        ReturnTwoDimes=0;
    end
end

endmodule

```

**Comment by fraxb04** [comment]
@Elia-Belli 
in questo caso, in base a cosa definiamo la quantità di Stati, e quale FSM utilizzare??
ho provato ad ideare una FSM di Moore senza resto, ma credo sia errata:
![image](https://user-images.githubusercontent.com/122573871/214304839-2ad422c0-dd39-41de-a314-87cbf4457197.png)

**Comment by Elia-Belli** [comment]
> @Elia-Belli in questo caso, in base a cosa definiamo la quantità di Stati, e quale FSM utilizzare?? ho provato ad ideare una FSM di Moore senza resto, ma credo sia errata: ![image](https://user-images.githubusercontent.com/122573871/214304839-2ad422c0-dd39-41de-a314-87cbf4457197.png)

Dipende dal tipo di problema se scegliere Moore o Mealy, in questo caso dovendo restituire il resto se utilizzassimo Moore dovremmo aggiungere degli stati appositi che diano come output il resto corretto per ogni caso, siccome non è detto che entrando in un determinato stato il resto sia sempre lo stesso. Mentre utilizzando Mealy è possibile restituirlo direttamente durante la transizione senza creare nuovi stati.

Per quanto riguarda il numero di stati non è necessario conoscerlo all'inizio, ci si rende conto man man che si disegna il diagramma di transizione.

Nella tuo diagramma di transizione ad ogni stato dovrebbe essere associato l'output corrispondente, in questo caso (siccome non vuoi dare il resto) deve esserci solo 1 o 0 ad indicare se si riceve la Soda (se 1 si torna allo stato iniziale), inoltre ogni stato deve aver definita la transizione per ognuno dei 3 input (5,10,25) invece alcuni dei tuoi stati non le hanno tutte. Ti consiglio di rifare il tutto usando Mealy che è molto meno complicato per un problema del genere.

**Comment by fraxb04** [comment]
@Elia-Belli 
ok grazie, but definendo per la transizione dei 3 input per ogni stato non si avrebbero più di 5 stati??

**Comment by Elia-Belli** [comment]
> @Elia-Belli ok grazie, but definendo per la transizione dei 3 input per ogni stato non si avrebbero più di 5 stati??

Utilizzando Mealy l'ho pensata così:
ho creato lo stato iniziale ed uno stato per ogni possibile quantità di soldi minore di 25, ovvero per 5, 10, 15, 20, ognuno di questi stati ha tutti gli output pari a 0 perchè la quantità è troppo piccola per acquistare la Soda, quando la quantità è $\geq 25$ si da la Soda e tornando allo stato iniziale si da il resto attivando i giusti output in ogni caso.
Quindi ho solo 5 stati

**Comment by fraxb04** [comment]
@Elia-Belli 
ok, thanks
